apps/web/src/app/(private)/projects/[projectId]/versions/[commitUuid]/documents/[documentUuid]/_components/DocumentEditor/Editor/Playground/Chat.tsx

```diff
@@ -0,0 +1,321 @@
+import { useCallback, useContext, useEffect, useRef, useState } from 'react'
+
+import {
+  Chain,
+  CompileError,
+  ContentType,
+  Conversation,
+  Message as ConversationMessage,
+  ConversationMetadata,
+  MessageRole,
+} from '@latitude-data/compiler'
+import {
+  ChainEventTypes,
+  LATITUDE_EVENT,
+  PROVIDER_EVENT,
+  type ChainEvent,
+  type DocumentVersion,
+} from '@latitude-data/core/browser'
+import { Text } from '$ui/ds/atoms'
+import { ErrorMessage, Message, MessageList } from '$ui/ds/molecules'
+import { ChatTextArea } from '$ui/ds/molecules/Chat/ChatTextArea'
+import { useAutoScroll } from '$ui/lib/hooks/useAutoScroll'
+import { cn } from '$ui/lib/utils'
+import { readStreamableValue } from 'ai/rsc'
+
+import { DocumentEditorContext } from '..'
+
+export default function Chat({
+  document,
+  metadata,
+  parameters,
+}: {
+  document: DocumentVersion
+  metadata: ConversationMetadata
+  parameters: Record<string, unknown>
+}) {
+  const { runAction, streamTextAction } = useContext(DocumentEditorContext)!
+  const [initialMetadata] = useState<ConversationMetadata>(metadata)
+  const [error, setError] = useState<Error | undefined>()
+  const [tokens, setTokens] = useState<number>(0)
+  const [isScrolledToBottom, setIsScrolledToBottom] = useState(false)
+  const containerRef = useRef<HTMLDivElement>(null)
+  useAutoScroll(containerRef, {
+    startAtBottom: true,
+    onScrollChange: setIsScrolledToBottom,
+  })
+
+  const [chain] = useState<Chain>(
+    new Chain({
+      prompt: initialMetadata.resolvedPrompt,
+      parameters,
+    }),
+  )
+  const runChainOnce = useRef(false)
+  const [chainLength, setChainLength] = useState<number>(Infinity) // Index where the chain ends and the chat begins
+  const [conversation, setConversation] = useState<Conversation | undefined>()
+  const [responseStream, setResponseStream] = useState<string | undefined>()
+
+  const addMessage = useCallback((message: ConversationMessage) => {
+    let newConversation: Conversation
+    setConversation((prevConversation) => {
+      newConversation = {
+        ...prevConversation,
+        messages: [...(prevConversation?.messages ?? []), message],
```

Comment:
Here maybe is more performant inserting the message in another way other than spread

Reply:
N is mega small and this is actually not used very often, not need to worry about performance here.